comp110-worksheet-6/OxoBoard.cs

```diff
@@ -6,53 +6,272 @@
 		// If the square is not empty, leave it as-is and return False.
 		public bool SetSquare(int x, int y, Mark mark)
 		{
-			throw new NotImplementedException("TODO: implement this function and then remove this exception");
+			if (board[x, y] == Mark.None)
+			{
+				board[x, y] = mark;
+				return true;
+
+			} 
+			else 
+			{
+				return false;
+			}
+
 		}
 
 		// If there are still empty squares on the board, return false.
 		// If there are no empty squares, return true.
 		public bool IsBoardFull()
 		{
-			throw new NotImplementedException("TODO: implement this function and then remove this exception");
+			for (int x = 0; x < board.GetLength(0); x++)
+			{
+				for (int y = 0; y < board.GetLength(1); y++)
+				{
+					if (board[x, y] == Mark.None) 
+					{
+						return false;
+					}
+				}
+			}
+
+			return true;
 		}
 
-		// If a player has three in a row, return Mark.O or Mark.X depending on which player.
+        private bool CheckDiagonals(int x, int y, Mark targetMark)
+        {
+            bool checkLeft = true;
+            bool checkRight = true;
+            bool winCondition = false;
+            // Check down and right diagonals.
+            // Only need to check downwards diagonals to still check every possibility.
+            int testX = x;
+            int testY = y;
+            testX++;
+            testY++;
+            if (testX < board.GetLength(0) && testY < board.GetLength(1))
+            {
+                if (board[testX, testY] == targetMark)
+                {
+                    int symbolCount = 2;
+                    do
+                    {
+                        testX++;
+                        testY++;
+                        if (testX < board.GetLength(0) && testY < board.GetLength(1))
+                        {
+                            if (board[testX, testY] == targetMark)
+                            {
+                                symbolCount++;
+                            } 
+                            else
+                            {
+                                checkRight = false;
+                            }
+
+                            if (symbolCount == itemsInARow)
+                            {
+                                winCondition = true;
+                            }
+                        }
+                        else
+                        {
+                            checkRight = false;
+                        }
+
+                    } while (checkRight && !winCondition);
+                }
+            }
+            
+
+            // If that was a win we don't need to check down and left diagonals so return the win.
+            if (winCondition)
+            {
+                return winCondition;
+            }
+
+            // Otherwise we check the down and left diagonals.
+            testX = x;
+            testY = y;
+
+            testX--;
+            testY++;
+
+            if (testX > 0 && testY < board.GetLength(1))
```

Comment:
This is a little repetitive -- could the two diagonals be checked using the same code (e.g. passing the direction as a parameter)?